builtin/logical/pki/cert_util.go

```diff
@@ -1513,3 +1513,288 @@ func stringToOid(in string) (asn1.ObjectIdentifier, error) {
+		// certificate itself.
+		referenceCert.CAChain = []string{referenceCert.Certificate}
+		return writeIssuer(ctx, s, referenceCert)
+	}
+
+	// Our provided reference cert might not be in the list of issuers. In
+	// that case, add it manually.
+	if referenceCert != nil {
+		missing := true
+		for _, issuer := range issuers {
+			if issuer == referenceCert.ID {
+				missing = false
+				break
+			}
+		}
+
+		if missing {
+			issuers = append(issuers, referenceCert.ID)
+		}
+	}
+
+	// We expect each of these maps to be the size of the number of issuers
+	// we have (as we're mapping from issuers to other values).
+	//
+	// The first caches the storage entry for the issuer, the second caches
+	// the parsed *x509.Certificate of the issuer itself, and the third maps
+	// that certificate back to the other issuers with that subject (note the
+	// keyword _other_: we'll exclude self-loops here).
+	issuerIdEntryMap := make(map[issuerId]*issuer, len(issuers))
+	issuerIdCertMap := make(map[issuerId]*x509.Certificate, len(issuers))
+	issuerIdParentsMap := make(map[issuerId][]issuerId, len(issuers))
+	issuerIdChildrenMap := make(map[issuerId][]issuerId, len(issuers))
+
+	// For every known issuer, we map that subject back to the id of issuers
+	// containing that subject. This lets us build our issuerId -> parents
+	// mapping efficiently. Worst case we'll have a single linear chain where
+	// every entry has a distinct subject.
+	subjectIssuerIdsMap := make(map[string][]issuerId, len(issuers))
+
+	// First, ready every storage entry from disk. We'll propagate entries to
+	// three of the maps here: all but issuerIdParentsMap and
+	// issuerIdChildrenMap, which we'll do in a second pass.
+	for _, identifier := range issuers {
+		var stored *issuer
+
+		// When the reference issuer is provided and matches this identifier,
+		// prefer the updated reference copy instead.
+		if referenceCert != nil && identifier == referenceCert.ID {
+			stored = referenceCert
+		} else {
+			// Otherwise, fetch it from disk.
+			stored, err = fetchIssuerById(ctx, s, identifier)
+			if err != nil {
+				return fmt.Errorf("unable to fetch issuer %v to build chain: %v", identifier, err)
+			}
+		}
+
+		if stored == nil || len(stored.Certificate) == 0 {
+			return fmt.Errorf("bad issuer while building chain: missing certificate entry: %v", identifier)
+		}
+
+		issuerIdEntryMap[identifier] = stored
+		cert, err := stored.GetCertificate()
+		if err != nil {
+			return fmt.Errorf("unable to parse issuer %v to certificate to build chain: %v", identifier, err)
+		}
+
+		issuerIdCertMap[identifier] = cert
+		subjectIssuerIdsMap[string(cert.RawSubject)] = append(subjectIssuerIdsMap[string(cert.RawSubject)], identifier)
+	}
+
+	// Now that we have the subj->issuer map built, we can build the parent
+	// and child mappings. We iterate over all issuers and build it one step
+	// at a time.
+	for _, child := range issuers {
```

Comment:
This is worst case `O(n^2)` because all of the issuers could have the same name and be self-signed certs. That makes the chain building (below) fast.